hudi-flink/src/main/java/org/apache/hudi/sink/bulk/BulkInsertWriterHelper.java

```diff
@@ -0,0 +1,192 @@
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * Helper class for bulk insert used by Flink.
+ */
+public class BulkInsertWriterHelper {
+
+  private static final Logger LOG = LogManager.getLogger(BulkInsertWriterHelper.class);
+
+  private final String instantTime;
+  private final int taskPartitionId;
+  private final long taskId;
+  private final long taskEpochId;
+  private final HoodieTable hoodieTable;
+  private final HoodieWriteConfig writeConfig;
+  private final RowType rowType;
+  private final Boolean arePartitionRecordsSorted;
+  private final List<HoodieInternalWriteStatus> writeStatusList = new ArrayList<>();
+  private HoodieRowDataCreateHandle handle;
+  private String lastKnownPartitionPath = null;
+  private final String fileIdPrefix;
+  private int numFilesWritten = 0;
+  private final Map<String, HoodieRowDataCreateHandle> handles = new HashMap<>();
+  private final KeyGenerator keyGenerator;
+  private final Schema schema;
+  private final RowDataKeyGen keyGen;
+
+  public BulkInsertWriterHelper(Configuration conf, HoodieTable hoodieTable, HoodieWriteConfig writeConfig,
+                                String instantTime, int taskPartitionId, long taskId, long taskEpochId, RowType rowType,
+                                boolean arePartitionRecordsSorted) {
+    this.hoodieTable = hoodieTable;
+    this.writeConfig = writeConfig;
+    this.instantTime = instantTime;
+    this.taskPartitionId = taskPartitionId;
+    this.taskId = taskId;
+    this.taskEpochId = taskEpochId;
+    this.rowType = addMetadataFields(rowType); // patch up with metadata fields
+    this.arePartitionRecordsSorted = arePartitionRecordsSorted;
+    this.fileIdPrefix = UUID.randomUUID().toString();
+    try {
+      this.keyGenerator = HoodieAvroKeyGeneratorFactory.createKeyGenerator(new TypedProperties(writeConfig.getProps()));
+    } catch (IOException e) {
+      throw new HoodieException("Create key generator error", e);
+    }
+    this.schema = getWriteSchema(writeConfig);
+    this.keyGen = RowDataKeyGen.instance(conf, rowType);
+  }
+
+  /**
+   * Returns the write instant time.
+   */
+  public String getInstantTime() {
+    return this.instantTime;
+  }
+
+  /**
+   * Get the schema of the actual write.
+   */
+  private static Schema getWriteSchema(HoodieWriteConfig config) {
+    return new Schema.Parser().parse(config.getWriteSchema());
+  }
+
+  public void write(RowData record) throws IOException {
+    try {
+      String recordKey = keyGen.getRecordKey(record);
+      String partitionPath = keyGen.getPartitionPath(record);
+
+      if ((lastKnownPartitionPath == null) || !lastKnownPartitionPath.equals(partitionPath) || !handle.canWrite()) {
+        LOG.info("Creating new file for partition path " + partitionPath);
```

Comment:
It is possible to reuse file here, should move this log when create new HoodieRowDataCreateHandle instance?

Reply:
No, the input data should sort by partition path.